### ISSUE TITLE
Dnsdist, using ProxyProtocol IXFR zone transfer fails, error - failed while receiving responses: end of file.

### ISSUE THREAD

**Issue by ksuuk** [PowerDNS/pdns#14540]
- [ X] This is not a support question, I have read [about opensource](https://www.powerdns.com/opensource.html) and will send support questions to the IRC channel, [Github Discussions](https://github.com/PowerDNS/pdns/discussions/) or the mailing list.

 - Program: dnsdist
 - Issue type: Bug report

### Short description
Bind IXFR zone transfer fails - “Transfer completed: 0 messages, 0 records, 0 bytes, 0.001 secs (0 bytes/sec) (serial 0).”

### Environment
 - Operating system: Ubuntu 22.04.4 LTS
 - Software version: Dnsdist 1.9.6, Bind 9.20.0
 - Software source: ISC and PowerDNS repository.

### Steps to reproduce
Change any server master zone/record on primary server and execute command “rndc reload”.

### Expected behaviour
The zone transfer does not fail, contains the changed data, and the transferred zone is updated on the secondary server.

### Actual behaviour
Using:

```
newServer({address="127.0.0.1:5353", tcpFastOpen=true, maxCheckFailures=3, useProxyProtocol=true})
newServer({address="[::1]:5353", tcpFastOpen=true, maxCheckFailures=3, useProxyProtocol=true})
```

The secondary server's bind transfer log shows only:

```
30-Jul-2024 19:15:23.280 notify: info: client @0x7f93a67b7800 192.168.50.5#42964: received notify for zone 'testst.com
```

Using:

```
newServer({address="127.0.0.1:5353", tcpFastOpen=true, maxCheckFailures=3, useProxyProtocol=true, useClientSubnet=true})
newServer({address="[::1]:5353", tcpFastOpen=true, maxCheckFailures=3, useProxyProtocol=true, useClientSubnet=true})
```

The secondary server's bind transfer log shows a bit more:

```
30-Jul-2024 19:24:54.040 notify: info: client @0x7f93a67b5c00 192.168.50.5#42964: received notify for zone 'testst.com'
30-Jul-2024 19:24:54.040 xfer-in: info: zone testst.com/IN: Transfer started. xfer-in: info: zone testst.com/IN: Transfer started.
30-Jul-2024 19:24:54.040 xfer-in: debug 1: zone testst.com/IN: requesting IXFR from 192.168.50.5#53
30-Jul-2024 19:24:54.044 xfer-in: info: 0x7f93afe9d000: transfer of 'testst.com/IN' from 192.168.50.5#53: connected using 192.168.50.5#53
30-Jul-2024 19:24:54.044 xfer-in: error: 0x7f93afe9d000: transfer of 'testst.com/IN' from 192.168.50.5#53: failed while receiving responses: end of file
30-Jul-2024 19:24:54.044 xfer-in: debug 1: zone testst.com/IN: zone transfer finished: end of file
30-Jul-2024 19:24:54.044 xfer-in: info: 0x7f93afe9d000: transfer of 'testst.com/IN' from 192.168.50.5#53: Transfer status: end of file
30-Jul-2024 19:24:54.044 xfer-in: info: 0x7f93afe9d000: transfer of 'testst.com/IN' from 192.168.50.5#53: Transfer completed: 0 messages, 0 records, 0 bytes, 0.001 secs (0 bytes/sec) (serial 0)
```

But as you see, zone transfer still fails.

When I disable dnsdist on the primary server, and modify named.conf so that bind listen on port 53:

```
listen-on port { any; };
listen-on-v6  { any; };
```

IXFR works fine:

```
30-Jul-2024 19:54:24.660 notify: info: client @0x7f93a6af5c00 192.168.50.5#56751: received notify for zone 'testst.com'
30-Jul-2024 19:54:24.660 xfer-in: info: zone testst.com/IN: Transfer started.
30-Jul-2024 19:54:24.660 xfer-in: debug 1: zone testst.com/IN: requesting IXFR from 192.168.50.5#53
30-Jul-2024 19:54:24.664 xfer-in: info: 0x7f93a4e75000: transfer of 'testst.com/IN' from 192.168.50.5#53: connected using 192.168.50.5#53
30-Jul-2024 19:54:24.664 xfer-in: debug 1: zone testst.com/IN: zone transfer finished: success
30-Jul-2024 19:54:24.664 xfer-in: info: zone testst.com/IN: transferred serial 2023021016
30-Jul-2024 19:54:24.664 xfer-in: info: 0x7f93a4e75000: transfer of 'testst.com/IN' from 192.168.50.5#53: Transfer status: success
30-Jul-2024 19:54:24.664 xfer-in: info: 0x7f93a4e75000: transfer of 'testst.com/IN' from 192.168.50.5#53: Transfer completed: 1 messages, 14 records, 651 bytes, 0.001 secs (651000 bytes/sec) (serial 2023021016)
```

Seems dnsdist interferes with this traffic somehow?

named.conf contains proxy config:

```
listen-on port 5353 proxy plain { any; };
listen-on-v6 port 5353 proxy plain { any; };
allow-proxy { any; };
allow-proxy-on { any; };
```

dnsdist.conf is:

```
setACL({'0.0.0.0/0', '::/0'})
addLocal('x.x.x.x:53', { reusePort=true, tcpFastOpenQueueSize=100, enableProxyProtocol=true})
addLocal('[xxx::1]:53', { reusePort=true, tcpFastOpenQueueSize=100, enableProxyProtocol=true})
newServer({address="127.0.0.1:5353", tcpFastOpen=true, maxCheckFailures=3, useProxyProtocol=true, useClientSubnet=true})
newServer({address="[::1]:5353", tcpFastOpen=true, maxCheckFailures=3, useProxyProtocol=true, useClientSubnet=true})
setServerPolicy(leastOutstanding)
setSecurityPollSuffix("")
pc = newPacketCache(10000000, {maxTTL=86400, minTTL=60, temporaryFailureTTL=30, staleTTL=300, deferrableInsertLock=true, keepStaleData=true, maxNegativeTTL=3600, numberOfShards=20, parseECS=false, maximumEntrySize=4096})
getPool(""):setCache(pc)
setRingBuffersSize(1000000, 100)
setMaxTCPClientThreads(20)
setMaxUDPOutstanding(65535)
local secondaryServersACL = newNMG()
secondaryServersACL:addMask("x.x.x.x")
addAction(NetmaskGroupRule(secondaryServersACL), AllowAction())
addAction(OrRule({OpcodeRule(DNSOpcode.Notify), OpcodeRule(DNSOpcode.Update), QTypeRule(DNSQType.AXFR), QTypeRule(DNSQType.IXFR)}), RCodeAction(DNSRCode.REFUSED))
addAction(QTypeRule(DNSQType.ANY), RCodeAction(DNSRCode.NOERROR))
addAction(AndRule{MaxQPSIPRule(5, 32, 64), TCPRule(false)}, TCAction())
addAction(AndRule{MaxQPSIPRule(7, 32, 64), TCPRule(true)}, DelayAction(100))
addAction(AndRule{MaxQPSIPRule(9, 32, 64), TCPRule(true)}, DropAction())
addAction(AllRule(), SetNoRecurseAction())
controlSocket("127.0.0.1")
setKey('xx')
webserver("0.0.0.0:80")
setWebserverConfig({password="xx", apiKey="xx", acl="0.0.0.0/0"})
```

**Comment by Habbie** [comment]
does IXFRing from BIND with `dig +proxy` work?

**Comment by ksuuk** [comment]
@Habbie 

If I use the command:
dig @x.x.x.x dummy.com SOA +noall +answer

to get zone serial nr, and then try to get zone incremental changes with command:
dig @x.x.x.x dummy.com IXFR=202408052 +multiline

then I only get SOA values, but no zone changes.

**Comment by Habbie** [comment]
> * [ X] I have read and understood the ['out in the open' support policy](https://blog.powerdns.com/2016/01/18/open-source-support-out-in-the-open/)

Please go read this again. Then update your text.

**Comment by ksuuk** [comment]
@Habbie Ok, I removed that part, please explain what should I update in my original configuration?

**Comment by Habbie** [comment]
In your original configuration, and in your comment above, you have hidden IPs and domain names. Don't do that.

**Comment by ksuuk** [comment]
@Habbie I'm sorry, but since my employer is already under constant DNS DDOS attacks, I'd rather not reveal the details, besides, they don't help solve the problem, and they don't change the fact that the dnsdist is having a bug.

**Comment by Habbie** [comment]
I will have to close this ticket until we get real configs and real program outputs.

**Comment by rgacogne** [comment]
Just to clarify, we don't need to know your public IP addresses. What we don't want is to have to guess at what your configuration really is, by trying to figure out what the IP addresses you redacted were, in order to reproduce the issue. If you can share a configuration with rfc-1918 addresses that we can use to reproduce the problem, that's fine and we will then of course fix any issue, but we are not OK with wasting our time with a redacted configuration that we cannot use directly.

**Comment by ksuuk** [comment]
@rgacogne

I'm sorry if I wasted your time, thanks for another chance.

For clarity, I've shared the exact configuration and only masked the public IP address of my primary authoritative "addLocal" and the "secondaryServersACL" of my secondary authoritative DNS server. You see, Dnsdist and Bind both run on the same server, so Dnsdist listens on port 53 and Bind listens on port 5353, and Dnsdist forwards requests to Bind using localhost. Note - since using ProxyProtocol Bind sees the client src IP address, so "allow-transfer" and "allow-notify" IP -s are configured in named.conf, so the secondary server ACL is not needed for security and I only use it for statistics.

So the simplest configuration of an authoritative master server for You might be:

```
setACL({'0.0.0.0/0', '::/0'})
addLocal('my public ipv4 address:53', { reusePort=true, tcpFastOpenQueueSize=100, enableProxyProtocol=true})
newServer({address="127.0.0.1:5353", tcpFastOpen=true, maxCheckFailures=3, useProxyProtocol=true, useClientSubnet=true})
setSecurityPollSuffix("")
```

BTW, as I mentioned, using "useClientSubnet" helped to get a bit more response data now, but I don't think it should be needed/used at all, when using ProxyProtocol.

**Comment by rgacogne** [comment]
So I tried reproducing with a bind 9.20.1 primary:
```
options {
    directory "/var/named";
    pid-file "/run/named/named.pid";

    allow-recursion { 127.0.0.1; };
    allow-transfer { 127.0.0.1; };
    allow-update { none; };

    version none;
    hostname none;
    server-id none;
    listen-on port 5353 proxy plain { any; };
    listen-on-v6 port 5353 proxy plain { any; };
    allow-proxy { any; };
    allow-proxy-on { any; };
};

zone "localhost" IN {
    type master;
    file "localhost.zone";
};
```
with dnsdist 1.9.6 in front:
```
setVerbose(true)
addLocal('127.0.0.1:8053', { reusePort=true, tcpFastOpenQueueSize=100, enableProxyProtocol=true})
newServer({address="127.0.0.1:5353", tcpFastOpen=true, maxCheckFailures=3, useProxyProtocol=true, useClientSubnet=true})
```

I can query the bind server directly:
```
$ dig IXFR=1 localhost @127.1 -p 5353 +proxy

; <<>> DiG 9.20.1 <<>> IXFR=1 localhost @127.1 -p 5353 +proxy
;; global options: +cmd
localhost.		86400	IN	SOA	localhost. root.localhost. 42 10800 900 604800 86400
localhost.		86400	IN	NS	localhost.
localhost.		86400	IN	A	127.0.0.1
localhost.		86400	IN	AAAA	::1
localhost.		86400	IN	SOA	localhost. root.localhost. 42 10800 900 604800 86400
;; Query time: 0 msec
;; SERVER: 127.0.0.1#5353(127.1) (TCP)
;; CLIENT PROXY HEADER: LOCAL
;; WHEN: Fri Aug 23 11:13:39 CEST 2024
;; XFR size: 5 records (messages 1, bytes 201)
```

and via dnsdist:
```
$ dig IXFR=1 localhost @127.1 -p 8053

; <<>> DiG 9.20.1 <<>> IXFR=1 localhost @127.1 -p 8053
;; global options: +cmd
localhost.		86400	IN	SOA	localhost. root.localhost. 42 10800 900 604800 86400
localhost.		86400	IN	NS	localhost.
localhost.		86400	IN	A	127.0.0.1
localhost.		86400	IN	AAAA	::1
localhost.		86400	IN	SOA	localhost. root.localhost. 42 10800 900 604800 86400
;; Query time: 0 msec
;; SERVER: 127.0.0.1#8053(127.1) (TCP)
;; WHEN: Fri Aug 23 11:13:34 CEST 2024
;; XFR size: 5 records (messages 1, bytes 212)
```

So I'm afraid I cannot reproduce the issue :-/ I have not tried adding a bind secondary because you did not share the corresponding configuration. Would you mind sharing it, and also letting me know if the `dig` commands I posted above (adapting the zone, IP and ports when needed, of course) yield similar results on your setup?

**Comment by ksuuk** [comment]
@rgacogne 

Thanks for testing, indeed I got the same result as you when testing, but everything still failed on the production servers using the very first configuration above. I tested for hours and came to the result that the cause is "newPacketCache".  Please try again to reproduce the problem with "newPacketCache" (in my test the other config options didn't matter).
```
cat /etc/dnsdist/dnsdist.conf
setACL({'0.0.0.0/0', '::/0'})
addLocal('x.x.x.x:53', { reusePort=true, tcpFastOpenQueueSize=100, enableProxyProtocol=true})
newServer({address="127.0.0.1:5353", tcpFastOpen=true, maxCheckFailures=3, useProxyProtocol=true})
setSecurityPollSuffix("")

pc = newPacketCache(10000000, {maxTTL=86400, minTTL=60, temporaryFailureTTL=30, staleTTL=300, deferrableInsertLock=true, keepStaleData=true,
   maxNegativeTTL=3600, numberOfShards=20, parseECS=false, maximumEntrySize=4096})
getPool(""):setCache(pc)
```
I used also named built in db.local for testing and queried from the same server, using it’s public IP (masked here):
```
cat /etc/bind/db.local
;
; BIND data file for local loopback interface
;
$TTL	604800
@	IN	SOA	localhost. root.localhost. (
			      5		; Serial
			 604800		; Refresh
			  86400		; Retry
			2419200		; Expire
			 604800 )	; Negative Cache TTL
;
@	IN	NS	localhost.
@	IN	A	127.0.0.1
@	IN	AAAA	::1
```
I changed the serial number from 4 to 5 + rndc reload

```
dig @x.x.x.x localhost SOA +noall +answer -p 53
;; Warning: Client COOKIE mismatch
localhost.		604779	IN	SOA	localhost. root.localhost. 4 604800 86400 2419200 604800

dig @x.x.x.x localhost SOA +noall +answer -p 5353 +proxy
localhost.		604800	IN	SOA	localhost. root.localhost. 5 604800 86400 2419200 604800
```
etc, it always failed.

Notice, that after restarting dnsdist the first IXFR request always succeeds and then all others fail.

Then I disabled the "newPacketCache".

```
cat dnsdist.conf
setACL({'0.0.0.0/0', '::/0'})
addLocal('x.x.x.x:53', { reusePort=true, tcpFastOpenQueueSize=100, enableProxyProtocol=true})
newServer({address="127.0.0.1:5353", tcpFastOpen=true, maxCheckFailures=3, useProxyProtocol=true})
setSecurityPollSuffix("")

--pc = newPacketCache(10000000, {maxTTL=86400, minTTL=60, temporaryFailureTTL=30, staleTTL=300, deferrableInsertLock=true, keepStaleData=true,
--    maxNegativeTTL=3600, numberOfShards=20, parseECS=false, maximumEntrySize=4096})
--getPool(""):setCache(pc)

dig @x.x.x.x localhost SOA +noall +answer -p 53
localhost.		604800	IN	SOA	localhost. root.localhost. 5 604800 86400 2419200 604800

dig @x.x.x.x localhost SOA +noall +answer -p 5353 +proxy
localhost.		604800	IN	SOA	localhost. root.localhost. 5 604800 86400 2419200 604800
```
I changed the serial number from 5 to 6 + rndc reload

```
dig @x.x.x.x localhost SOA +noall +answer -p 53
localhost.		604800	IN	SOA	localhost. root.localhost. 6 604800 86400 2419200 604800

dig @x.x.x.x localhost SOA +noall +answer -p 5353 +proxy
localhost.		604800	IN	SOA	localhost. root.localhost. 6 604800 86400 2419200 604800
```
etc, it always succeeded.

In summary, the IXFR request fails when using "newPacketCache", even according to https://dnsdist.org/advanced/axfr.html?highlight=axfr “Since 1.8.0, dnsdist will no longer cache responses to AXFR and IXFR queries.”

I tried rule:
```
addAction(AndRule({NetmaskGroupRule(secondaryServersACL), OrRule({QTypeRule(DNSQType.AXFR), QTypeRule(DNSQType.IXFR)})}), SetSkipCacheAction())
```

but this din’t help, the rule:
```
addAction(NetmaskGroupRule(secondaryServersACL), SetSkipCacheAction())
```
works, but this affects to all cache, when I need just skip AXFR and IXFR queries.

Please fix it.

**Comment by rgacogne** [comment]
`AXFR` and `IXFR` are not cached, I just made sure of that via testing and looking at the code (https://github.com/PowerDNS/pdns/blob/rel/dnsdist-1.9.x/pdns/dnsdist-cache.cc#L126). I'm wondering if the secondary does a `SOA` query before the `IXFR` one: if it does, the `SOA` response is likely cached. You can disable that with something like this:

```lua
addAction(AndRule{NetmaskGroupRule(secondaryServersACL), QTypeRule(DNSQType.SOA)}, SetSkipCacheAction())
```

dnsdist has no way of knowing whether `SOA` responses should be cached, or more exactly when they should be purged.

**Comment by ksuuk** [comment]
@rgacogne

It seems that the master sends with the notify a new serial, and that the based that, the secondary makes an IXFR query, but since it receives the old, cached serial in response, the zone transfer is not performed.

The rule You suggested works, but I recommend disabling the SOA caching directly in the code and/or adding information about it in the documentation.

**Comment by rgacogne** [comment]
> the secondary makes an IXFR query, but since it receives the old, cached serial in response, the zone transfer is not performed.

I believe there is a step missing there: the secondary likely makes a `SOA` query before a `IXFR` one.

>  recommend disabling the SOA caching directly in the code

I can see how it would make sense for your use case, but it would be very unexpected for other ones. Most users would not like `SOA` queries to just go through, as it would increase the load on the backend for no good reason.

> adding information about it in the documentation

Sure, if you can spare a few minutes to come up with a few sentences, or even better open a pull request, that would go a long way.

**Comment by rgacogne** [comment]
In the meantime I'm closing this issue since as far as I can tell there is no bug in dnsdist.

**Comment by ksuuk** [comment]
@rgacogne 

Thanks, maybe few notes to - https://dnsdist.org/advanced/axfr.html?highlight=axfr

-    -- this rule will not stop the processing, but disable caching for AXFR and IXFR responses
-addAction(OrRule({QTypeRule(DNSQType.AXFR), QTypeRule(DNSQType.IXFR)}), SetSkipCacheAction())

+    -- this rule will not stop the processing, but disable caching for AXFR, IXFR and SOA responses
+ addAction(OrRule({QTypeRule(DNSQType.AXFR), QTypeRule(DNSQType.IXFR, QTypeRule(DNSQType.SOA)}), SetSkipCacheAction())

-Changed in version 1.8.0: Since 1.8.0, dnsdist will no longer cache responses to AXFR and IXFR queries.
+Changed in version 1.8.0: Since 1.8.0, dnsdist will no longer cache responses to AXFR and IXFR queries, but still caches responses to SOA queries.
